### DISCUSSION TITLE
Typescript: Deno as LSP backend

### DISCUSSION THREAD

**Discussion by fosskers** [syl20bnr/spacemacs#14423] [Ideas]
So I just got [Deno](https://deno.land/) working as the LSP backend for the `typescript` layer. I only had to tweak my own config - luckily no deeper elisp alteration was necessary. I just added the following two variable changes to my `lsp` layer:
```elisp
     (lsp :variables
          lsp-clients-typescript-tls-path "/bin/deno"
          lsp-clients-typescript-server-args "lsp")
```
[`lsp-javascript`](https://github.com/emacs-lsp/lsp-mode/blob/e7b86d076b41e89254220122a1ea5b05f03cc4eb/clients/lsp-javascript.el) thinks it's running `typescript-language-server`, but the above config seems to be enough to trick it. 

I notice that a number of other language layers have the ability to (somewhat) trivially switch between LSP providers. Should we do the same for typescript?

**Comment by fosskers** [comment]
`deno` also supplies its own linter and formatter, so those would have to be hooked in too.

**Reply by kitsonk** [reply]
Formatting is just part of the LSP protocol, and the LSP responds to `textDocument/formatting`.  It doesn't currently support format on paste or format on type yet.  The formatter is very opinionated at the moment, and doesn't really respect any options, as it aligns to what occurs when you do `deno fmt` with the Deno CLI.

The linting is effectively automatic, if the configuration from the client is reported to have `deno.lint` configuration option set to `true` for a workspace/globally/etc.  It will simply come back as part of the diagnostics with the source of `deno-lint`.  Like formatting, it is highly opinionated and not configurable, which also aligns to the `deno lint` capability built into the Deno CLI.

**Comment by kitsonk** [comment]
👋  I work on the Deno LSP.  It has some strong opinions in it when it comes to how Deno handles JavaScript and TypeScript that _might_ not work as general purpose as some other integration.  Deno resolves modules as fully qualified module names with no "magic" resolution, where as the far more common Node.js has a fairly complex module resolution algorithm and extensions are optional (and quite uncommon).

In the internals of the LSP, we basically "drop" all the diagnostics related to module resolution and quite a few others that aren't meaningful to a Deno context that come back from the TypeScript compiler, and we replace the module resolution ones with our own.

Currently the Deno LSP is missing a few features that language servers provide.  The roadmap is here: https://github.com/denoland/deno/issues/8643.

**Reply by kitsonk** [reply]
Just to mention too that the source from the diagnostics are `deno-ts` for ones that come back from the integrated version of TypeScript/tsc in Deno and `deno` for the ones that come from our internal module resolution and caching code.

**Reply by fosskers** [reply]
Thanks. Let's keep this train of thought open. For now I'll keep using the "hacked" settings I have above and see how well it works. So far I'm happy. Linting seems to work, but I'm not sure if it's info from Deno, or the default `tide` linter at work within the `typescript` layer. Formatting is taken care of by `prettier`. For reference, here are my `typescript` layer settings:
```
     (typescript :variables
                 typescript-fmt-on-save t
                 typescript-fmt-tool 'prettier)
```
I tried to hack deno in there too for the `fmt-tool`, but it turns out the `'prettier` symbol there is just that - a symbol that unlocks a bunch of custom spacemacs code, and not a raw command line command. Using `deno fmt` would require some elisp to be written in the same way.

**Reply by simne7** [reply]
While trying to get the javascript layer and deno work together nicely I stumbled on the [deno-fmt](https://github.com/rclarey/deno-emacs) package. I followed [the install instructions](https://github.com/rclarey/deno-emacs?tab=readme-ov-file#spacemacs) and disabled the fmt-tool. E.g. in a `.dir-locals.el` 

```
((nil . ((lsp-enabled-clients . (deno-ls))))
 (js2-mode . ((javascript-fmt-tool . nil)
              ;; (after-save-hook . (lambda() (shell-command (format "deno fmt %s" buffer-file-name))))
              (javascript-backend . lsp))))
```

`deno-fmt` works for me so I could disable my workaround that formatted the file after saving.

**Comment by lebensterben** [comment]
Similar to the treatment of Node.js
https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Btools/node

Deno should have its own layer instead of being a part of Typescript.

**Comment by fosskers** [comment]
Apparently [lsp-deno](https://emacs-lsp.github.io/lsp-mode/page/lsp-deno/) exists!

**Comment by basus** [comment]
I'm having trouble getting Deno working as my Typescript backend. I'm running the latest `develop` branch and I have Deno installed. When installing the typescript layer, I'm not prompted to install an LSP, and the Deno LSP is started when I open a `.ts` file. But then I get the following message in the `*lsp-log*` buffer:

```
Command "deno lsp" is present on the path.
Command "semgrep lsp" is not present on the path.
Command "deno lsp" is present on the path.
Command "semgrep lsp" is not present on the path.
Found the following clients for <filename>: (server-id deno-ls, priority -5)
```

And in my `*Messages*` buffer I see a message:
```
Error processing message (error "The package typescript is not installed.  Unable to find nil").
```

Even though the Deno LSP is running and Spacemacs is connected to it, I only get timeout messages if I try any actual LSP actions. I can't tell what's doing on and I'm not sure how to go about debugging it. I'd appreciate any help.

**Reply by zhaozhixu** [reply]
I have the exactly same problem with emacs 29.4 + deno 1.46.3. @kitsonk Do you have any insight to this?

**Reply by zhaozhixu** [reply]
@basus @kitsonk  I looked up into lsp-mode code, and found that it was an error in `lsp--npm-dependency-path` . Somehow the lsp server cannot find the `npm` path.

Trying to `M-x lsp-install-server` then install `ts-ls` server can magically solve this problem. Although it seems to have nothing to do with `deno-ls` and I have to manually disable `ts-ls` after installing it by `(lsp-disabled-clients '(ts-ls))`.

**Reply by MrMugame** [reply]
Just installing typescript globally did it for me
`sudo npm install -g typescript`
